clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -233,7 +235,7 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, C
             // This is to prevent later batches from being expired while an earlier batch is still in progress.
             // Note that `muted` is only ever populated if `max.in.flight.request.per.connection=1` so this protection
             // is only active in this case. Otherwise the expiration order is not guaranteed.
-            if (!muted.contains(tp)) {
+            if (!muted.contains(tp) && (staleMetadataNow || cluster.leaderFor(tp) == null)) {
```

Comment:
Do we want to expire when the metadata is stale? or do we actually want to do otherwise?